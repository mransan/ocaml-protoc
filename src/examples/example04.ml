open Example04_types
open Example04_pp

let (@+) value next = 
  Cons {value;next}

let () = 
  let l = 1 @+ 2 @+ 3 @+ 4 @+ 5 @+ Nil in 
  Format.(fprintf std_formatter "l = %a\n" pp_int_list l)
