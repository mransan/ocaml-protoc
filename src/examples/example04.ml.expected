[@@@ocaml.warning "-27-30-39"]

type int_list_nil = unit

type int_list_cons = {
  value : int;
  next : int_list;
}

and int_list_t =
  | Cons of int_list_cons
  | Nil

and int_list = {
  t : int_list_t option;
}

let rec default_int_list_nil = ()

let rec default_int_list_cons 
  ?value:((value:int) = 0)
  ?next:((next:int_list) = default_int_list ())
  () : int_list_cons  = {
  value;
  next;
}

and default_int_list_t () : int_list_t = Cons (default_int_list_cons ())

and default_int_list 
  ?t:((t:int_list_t option) = None)
  () : int_list  = {
  t;
}

[@@@ocaml.warning "-27-30-39"]

(** {2 Formatters} *)

let rec pp_int_list_nil fmt (v:int_list_nil) = 
  let pp_i fmt () =
    Pbrt.Pp.pp_unit fmt ()
  in
  Pbrt.Pp.pp_brk pp_i fmt ()

let rec pp_int_list_cons fmt (v:int_list_cons) = 
  let pp_i fmt () =
    Pbrt.Pp.pp_record_field ~first:true "value" Pbrt.Pp.pp_int fmt v.value;
    Pbrt.Pp.pp_record_field ~first:false "next" pp_int_list fmt v.next;
  in
  Pbrt.Pp.pp_brk pp_i fmt ()

and pp_int_list_t fmt (v:int_list_t) =
  match v with
  | Cons x -> Format.fprintf fmt "@[<hv2>Cons(@,%a)@]" pp_int_list_cons x
  | Nil  -> Format.fprintf fmt "Nil"

and pp_int_list fmt (v:int_list) = 
  let pp_i fmt () =
    Pbrt.Pp.pp_record_field ~first:true "t" (Pbrt.Pp.pp_option pp_int_list_t) fmt v.t;
  in
  Pbrt.Pp.pp_brk pp_i fmt ()
